docker/docker_client_test.go

```diff
@@ -88,194 +88,215 @@ func TestDockerCertDir(t *testing.T) {
+			},
+			{
+				name:             "match one",
+				hostname:         "example.org",
+				authConfig:       makeTestAuthConfig(testAuthConfigDataMap{"example.org": testAuthConfigData{"joe", "mypass"}}),
+				expectedUsername: "joe",
+				expectedPassword: "mypass",
+			},
+			{
+				name:       "match none",
+				hostname:   "registry.example.org",
+				authConfig: makeTestAuthConfig(testAuthConfigDataMap{"example.org": testAuthConfigData{"joe", "mypass"}}),
+			},
+			{
+				name:     "match docker.io",
+				hostname: "docker.io",
+				authConfig: makeTestAuthConfig(testAuthConfigDataMap{
+					"example.org": testAuthConfigData{"example", "org"},
+					"docker.io":   testAuthConfigData{"docker", "io"},
+				}),
+				expectedUsername: "docker",
+				expectedPassword: "io",
+			},
+			{
+				name:     "match docker.io normalized",
+				hostname: "docker.io",
+				authConfig: makeTestAuthConfig(testAuthConfigDataMap{
+					"example.org":                testAuthConfigData{"bob", "pw"},
+					"https://index.docker.io/v1": testAuthConfigData{"alice", "wp"},
+				}),
+				expectedUsername: "alice",
+				expectedPassword: "wp",
+			},
+			{
+				name:     "normalize registry",
+				hostname: "https://docker.io/v1",
+				authConfig: makeTestAuthConfig(testAuthConfigDataMap{
+					"docker.io":      testAuthConfigData{"user", "pw"},
+					"localhost:5000": testAuthConfigData{"joe", "pass"},
+				}),
+				expectedUsername: "user",
+				expectedPassword: "pw",
+			},
+			{
+				name:     "match localhost",
+				hostname: "http://localhost",
+				authConfig: makeTestAuthConfig(testAuthConfigDataMap{
+					"docker.io":   testAuthConfigData{"user", "pw"},
+					"localhost":   testAuthConfigData{"joe", "pass"},
+					"example.com": testAuthConfigData{"alice", "pwd"},
+				}),
+				expectedUsername: "joe",
+				expectedPassword: "pass",
+			},
+			{
+				name:     "match ip",
+				hostname: "10.10.3.56:5000",
+				authConfig: makeTestAuthConfig(testAuthConfigDataMap{
+					"10.10.30.45":     testAuthConfigData{"user", "pw"},
+					"localhost":       testAuthConfigData{"joe", "pass"},
+					"10.10.3.56":      testAuthConfigData{"alice", "pwd"},
+					"10.10.3.56:5000": testAuthConfigData{"me", "mine"},
+				}),
+				expectedUsername: "me",
+				expectedPassword: "mine",
+			},
+			{
+				name:     "match port",
+				hostname: "https://localhost:5000",
+				authConfig: makeTestAuthConfig(testAuthConfigDataMap{
+					"https://127.0.0.1:5000": testAuthConfigData{"user", "pw"},
+					"http://localhost":       testAuthConfigData{"joe", "pass"},
+					"https://localhost:5001": testAuthConfigData{"alice", "pwd"},
+					"localhost:5000":         testAuthConfigData{"me", "mine"},
+				}),
+				expectedUsername: "me",
+				expectedPassword: "mine",
+			},
+			{
+				name:     "use system context",
+				hostname: "myexample.org",
+				authConfig: makeTestAuthConfig(testAuthConfigDataMap{
+					"myexample.org": testAuthConfigData{"user", "pw"},
```

Comment:
This used to be `example.org`; is the change to `myexample` important?

Reply:
So when it was example.org, the `match one` test case kept failing as it would get the username and password from this test instead. I really couldn't understand why that was happening...
I changed the hostname, and it was fine.
Do you have any idea why that might be happening?

Reply:
Nothing jumps at me as a likely reason, but that does seems like something worth investigating.  

If the test cases affect each other that way, that makes the reliability of all of them questionable, especially considering how they like to reuse the same user/password combinations.

Reply:
This failure is only happening when the config path is `~/.docker/config.json` though.

Reply:
OK, that explains it – the first iteration top-level loop creates writes data to `auth.json`, and leaves the last version of it around. The second iteration of the top-level loop is checking for data in `auth.json` first, and only if it is not found there, in `.docker/config.json`; IOW the surviving `auth.json` is affecting the behavior.

`os.Remove(configPath)` at the end of the outer loop body should fix this.

Reply:
Right, I didn't even think of that. Thanks! :D Fixed